index.rst

```diff
@@ -1,29 +1,125 @@
+
+There should be two versions of each of these routes, one that returns HTML for use by humans and one that returns JSON for use by automated tests (see :ref:`automated-testing`).
+
+.. _automated-testing:
+
+Automated testing
+=================
+
+Use cases
+---------
+
+In addition to allowing human verification of the basic functionality of a new deployment, we also want to support automated testing.
+This will come in two forms:
+
+- Optionally run a validation test of a newly-deployed cluster immediately after deploying it.
+  This would be done in conjunction with asking Argo CD if health checks are passing for all applications and would be done in the installer.
+  The CI minikube deployment would run this check.
+
+- Regularly run the same validation tests for existing clusters so that we know if any of the basic functionality breaks.
+  This is well-suited for mobu, which already runs similar checks for existing clusters.
+
+Proposed implementation
+-----------------------
+
+The JSON routes of Muster should, in addition to summarizing all of the checks run and any errors, provide a top-level pass-fail status.
+We can then write a Muster client that does the following:
+
+#. Obtain the Gafaelfawr bootstrap token from the Gafaelfawr secret.
+#. Using that token, create a token for a ``bot-check-muster`` user with ``exec:admin`` scope.
+#. Query all of the Muster JSON endpoints, providing that token to the ones requiring authentication, and check that the top-level status is pass.
+#. Query a Muster endpoint that requires authentication without providing the token and check that a 401 status is returned.
+#. Create a token for ``bot-check-muster`` without any scopes, and query one of the Muster endpoints that requires authentication.
+   Check the return status and the correctness of the headers.
+#. Pass an invalid token to the same endpoint and check the return status and correctness of the headers and response body.
+
+The last two steps test the special Gafaelfawr integration with ingress-nginx to pass additional headers and a correct return status and response body when an auth subrequest handler fails.
+
+For ongoing testing, we want mobu to run the same code periodically.
+
+To avoid duplicating the same code in multiple places, we can add this code to mobu as a new ``MusterRunner`` class.
+Then, add a new endpoint to mobu that, rather than starting a continuous runner, executes a runner once and returns its results directly as the response of that endpoint.
+Validation testing can then be done by deploying mobu without any configured monkeys (test runners), and then invoking that endpoint to run the ``MusterRunner``.
+(The drawback of this approach is that it adds mobu to the minimum application set for a Phalanx deployment, but that seems better than duplicating this code or creating a new special-purpose library package that will require ongoing maintenance.)
```

Comment:
I don't see a problem with that, mobu is fairly lightweight, does not require resources that would not be otherwise available to phalanx and so on.